### PR TITLE
[full-ci][gui-tests] Automated GUI tests in windows

### PR DESCRIPTION
**Pull request by saw-jan**
Part of https://github.com/owncloud/client/issues/9833

From the initial work on squish tests for Windows, it looks like the current tests can be run in Windows as well with some tweaking.
Almost all the tests use the sync pattern matching to determine whether the sync has been successful or not. In Linux, we utilize socket connection and Windows has named pipe for communication.

Tests skipped for Windows:
- `Sharing link with expiration date` (difference in date format)
- `Syncing a folder having space at the end` (white space is trimmed)
- `Filenames that are rejected by the server are reported` (invalid file name for windows)
- `Sync long nested folder` (windows doesn't allow long path name)
- `Invalid system names are synced in linux` (linux only)

Feature disabled for Windows:
- Screenshots
- Stacktracing

Tasks:
- [x] adjust tests for Windows
- [x] run tests with oc10 server
- [ ] run tests with ocis server (problem with cleanup)

### CODE REVIEW

File: test/gui/shared/scripts/helpers/WinPipeHelper.py [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+                _, message_mem = win32file.ReadFile(self._pipe, DEFAULT_BUFLEN, self._overlapped)
+                if message_mem:
+                    m_bytes = bytes(message_mem)
+                    if b'\n' in m_bytes:
+                        for m in m_bytes.split(b'\n'):
+                            print(type(m))
+                            try:
+                                # append decodable bytes
+                                m.decode('utf-8')
+                                messages += m + b'\n'
```
Comment (saw-jan) [inline test/gui/shared/scripts/helpers/WinPipeHelper.py:90]: filter out the unwanted messages

**Comment by TheOneRing** [comment]
> @TheOneRing what does this error mean?
> 
> Screenshot from 2023-11-30 16-19-08
> 
> [client.log](https://github.com/owncloud/client/files/13511685/client.log)

It usually means that a space that we synced was removed/disabled on the server.

File: test/gui/shared/scripts/helpers/ConfigHelper.py [hunk 2]
```diff
@@ -49,14 +84,12 @@ def get_config_from_env_file(env):
...
-        get_config_from_env_file("XDG_CONFIG_HOME"), "ownCloud"
-    ),
-    'guiTestReportDir': os.path.abspath('../reports/'),
+    'clientRootSyncPath':  getClientRootPath(),
+    'tempFolderPath':  path.join(getClientRootPath(), 'temp'),
+    'clientConfigDir': getConfigHome(),
+    'guiTestReportDir': path.abspath(path.join('..', 'reports')),
     'ocis': False,
-    'custom_lib': os.path.abspath('../shared/scripts/custom_lib'),
+    'custom_lib': path.abspath(path.join('..', 'shared', 'scripts', 'custom_lib')),
```
Comment (TheOneRing) [inline test/gui/shared/scripts/helpers/ConfigHelper.py:92]: a more modern approach for paths is 
```
from pathlib import Path
foo = Path("..") / "shared/scrips/custom_lib"
or
foo = Path("../shared/scrips/custom_lib")
```
But actually 90% of windows accepts forwards slashes in paths.

File: test/gui/shared/scripts/helpers/SetupClientHelper.py [hunk 3]
```diff
@@ -107,29 +108,29 @@ def getPollingInterval():
...
-    0/dav_user={davUserName}
-    0/display-name={displayUserName}
-    0/http_CredentialVersion=1
-    0/http_oauth={oauth}
-    0/http_user={davUserName}
-    0/url={local_server}
-    0/user={displayUserFirstName}
-    0/version=1
+    0{path_sep}Folders{path_sep}1{path_sep}davUrl={url}
+    0{path_sep}Folders{path_sep}1{path_sep}ignoreHiddenFiles=true
```
[the remark this answers is not in the document]
Comment (saw-jan) [reply test/gui/shared/scripts/helpers/SetupClientHelper.py:112]: config in windows has `\` so I was trying to make similar to it

File: test/gui/shared/scripts/helpers/WinPipeHelper.py [hunk 4]
```diff
@@ -0,0 +1,133 @@
+import os
```
Comment (saw-jan) [inline test/gui/shared/scripts/helpers/WinPipeHelper.py:1]: @TheOneRing Could you review this PR? Specially this file

File: test/gui/shared/scripts/helpers/SyncHelper.py [hunk 5]
```diff
@@ -63,7 +66,23 @@
         # when syncing an account that has some files/folders
         [SYNC_STATUS['SYNC'], SYNC_STATUS['OK']],
     ],
-    'synced': [SYNC_STATUS['SYNC'], SYNC_STATUS['OK']],
+    'synced': [
+        [
+            SYNC_STATUS['SYNC'],
```
Comment (TheOneRing) [inline test/gui/shared/scripts/helpers/SyncHelper.py:71]: This looks a tiny bit weird?

File: test/gui/shared/scripts/helpers/WinPipeHelper.py [hunk 6]
```diff
@@ -0,0 +1,133 @@
...
+import time
+import win32pipe, win32file, winerror, pywintypes, win32event
+
+TIMEOUT = 100
+DEFAULT_BUFLEN = 4096
+CLIENT_MESSAGES = [
+    'STATUS',
```
Comment (TheOneRing) [inline test/gui/shared/scripts/helpers/WinPipeHelper.py:8]: They are the same on all platforms maybe we can share them?

File: test/gui/shared/scripts/helpers/WinPipeHelper.py (hunk 4 above)
Comment (TheOneRing) [reply test/gui/shared/scripts/helpers/WinPipeHelper.py:1]: Looks good.

File: test/gui/shared/scripts/helpers/WinPipeHelper.py (hunk 6 above)
Comment (saw-jan) [reply test/gui/shared/scripts/helpers/WinPipeHelper.py:8]: yeah, but as of now I don't see other usages. we don't need to filter out socket messages in linux and also the socket connection helper file is from https://raw.github.com/owncloud/client-desktop-shell-integration-nautilus/master/src/syncstate.py, so I won't dare to edit it :sweat_smile:

**Comment by saw-jan** [comment]
> > * `Sharing link with expiration date` (difference in date format)
> 
> How do they differ?
> 
> > * `Syncing a folder having space at the end` (white space is trimmed)
> > * `Filenames that are rejected by the server are reported` (invalid file name for windows)
> > * `Sync long nested folder` (windows doesn't allow long path name)
> 
> Hmmm https://learn.microsoft.com/en-us/windows/win32/fileio/maximum-file-path-limitation?tabs=registry#enable-long-paths-in-windows-10-version-1607-and-later or https://stackoverflow.com/questions/1365797/python-long-filename-support-broken-in-windows
> 
> > * `Invalid system names are synced in linux` (linux only)

Thanks for the useful links :smiley:

I would like to follow up on these tests after this PR along with the test cleanup while testing against the ocis server.

**Review by TheOneRing — APPROVED** [review]

**Pull request merged**